addons/web/controllers/main.py

```diff
@@ -990,140 +990,183 @@ def action(self, model, id):
 
-    def placeholder(self, image='placeholder.png'):
-        addons_path = http.addons_manifest['web']['addons_path']
-        return open(os.path.join(addons_path, 'web', 'static', 'src', 'img', image), 'rb').read()
+    # get content after cache control
+    if status == 200 and (xmlid or id):
+        content = getattr(obj, field)
+        if not content and getattr(obj, 'url', None):
+            return (302, headers, obj.url)
 
-    @http.route('/web/binary/saveas', type='http', auth="public")
-    @serialize_exception
-    def saveas(self, model, field, id=None, filename_field=None, **kw):
-        """ Download link for files stored as binary fields.
-
-        If the ``id`` parameter is omitted, fetches the default value for the
-        binary field (via ``default_get``), otherwise fetches the field for
-        that precise record.
-
-        :param str model: name of the model to fetch the binary from
-        :param str field: binary field
-        :param str id: id of the record from which to fetch the binary
-        :param str filename_field: field holding the file's name, if any
-        :returns: :class:`werkzeug.wrappers.Response`
-        """
-        Model = request.registry[model]
-        cr, uid, context = request.cr, request.uid, request.context
-        fields = [field]
-        content_type = 'application/octet-stream'
-        if filename_field:
-            fields.append(filename_field)
-        if id:
-            fields.append('file_type')
-            res = Model.read(cr, uid, [int(id)], fields, context)[0]
-            if res.get('file_type'):
-                content_type = res['file_type']
-        else:
-            res = Model.default_get(cr, uid, fields, context)
-        filecontent = base64.b64decode(res.get(field) or '')
-        if not filecontent:
-            return request.not_found()
-        else:
-            filename = '%s_%s' % (model.replace('.', '_'), id)
-            if filename_field:
-                filename = res.get(filename_field, '') or filename
-            return request.make_response(filecontent,
-                [('Content-Type', content_type),
-                 ('Content-Disposition', content_disposition(filename))])
-
-    @http.route('/web/binary/saveas_ajax', type='http', auth="public")
-    @serialize_exception
-    def saveas_ajax(self, data, token):
-        jdata = simplejson.loads(data)
-        model = jdata['model']
-        field = jdata['field']
-        data = jdata['data']
-        id = jdata.get('id', None)
-        filename_field = jdata.get('filename_field', None)
-        context = jdata.get('context', {})
-        content_type = 'application/octet-stream'
+    # return
+    return (status, headers, content)
 
-        Model = request.session.model(model)
-        fields = [field]
-        if filename_field:
-            fields.append(filename_field)
+class Binary(http.Controller):
+
+    # backward compatibility
```

Comment:
Why?
If code does not call them anymore, no need to have these routes (they are implementation details)